alf/experience_replayers/segment_tree.py

```diff
@@ -0,0 +1,216 @@
+            right = self._values[indices * 2 + 1]
+            self._values[indices] = op(left, right)
+            return indices
+
+        with alf.device(self._device):
+            indices = convert_device(indices)
+            values = convert_device(values)
+
+            assert indices.ndim == 1
+            assert values.ndim == 1
+            assert indices.shape == values.shape, (
+                "indices and values should be 1-D tensor with the length. Got "
+                "%s and %s." % (indices.shape, values.shape))
+            op = self._op
+            indices, order = torch.sort(indices)
+            values = values[order]
+            assert indices[-1] < self._capacity
+            indices = self._index_to_leaf(indices)
+            self._values[indices] = values
+
+            num_large = (indices >= self._leftest_leaf).to(torch.int64).sum()
+            if num_large > 0:
+                large_indices = indices[:num_large]
+                small_indices = indices[num_large:]
+                large_indices = _step(large_indices)
+                indices = torch.cat([large_indices, small_indices])
+
+            for _ in range(self._depth):
+                indices = _step(indices)
+
+    def __getitem__(self, idx):
+        """Get the values of leaves.
+
+        Args:
+            idx (Tensor): 1-D int64 Tensor. Its values should be in range
+                [0, capacity).
+        Returns:
+            Tensor: with same shaps as idx.
+        """
+        with alf.device(self._device):
+            idx = convert_device(idx)
+            assert 0 <= idx.min()
+            assert idx.max() < self._capacity
+            result = self._values[self._index_to_leaf(idx)]
+        return convert_device(result)
+
+    def _index_to_leaf(self, idx):
+        """
+        Make sure idx=0 is the leftest leaf.
+        """
+        idx += self._leftest_leaf
+        idx = torch.where(idx >= 2 * self._capacity, idx - self._capacity, idx)
+        return idx
+
+    def _leaf_to_index(self, leaf):
+        idx = leaf - self._leftest_leaf
+        idx = torch.where(idx < 0, idx + self._capacity, idx)
+        return idx
+
+    def summary(self):
+        """The summary of the tree.
+
+        If ``op`` is ``torch.add``, it's the sum of all values.
+        If ``op`` is ``torch.min``, it's the min of all values.
+        If ``op`` is ``torch.max``, it's the max of all values.
+
+        Returns:
+            a scalar
+        """
+        return convert_device(self._values[1])
+
+
+class SumSegmentTree(SegmentTree):
+    """SegmentTree with sum operation."""
+
+    def __init__(self,
+                 capacity,
+                 dtype=torch.float32,
+                 device="cpu",
+                 name="SumSegmentTree"):
+        super().__init__(
+            capacity, torch.add, dtype=dtype, device=device, name=name)
+
+    def __setitem__(self, indices, values):
+        assert values.min() >= 0
```

Comment:
should be > 0 instead so that find_sum_bound has a unique solution?

Reply:
0 is okay. find_sum_bound() will never return an index with zero value.